benchmark/Streamly/Benchmark/Prelude/Serial/Transformation2.hs

```diff
@@ -0,0 +1,425 @@
+    -> Int
+    -> SerialT m Int
+iterateSingleton g count n = iterateN g (return n) count
+
+-- XXX need to check why this is slower than the explicit recursion above, even
+-- if the above code is written in a foldr like head recursive way. We also
+-- need to try this with foldlM' once #150 is fixed.
+-- However, it is perhaps best to keep the iteration benchmarks independent of
+-- foldrM and any related fusion issues.
+{-# INLINE _iterateSingleton #-}
+_iterateSingleton ::
+       S.MonadAsync m
+    => (Int -> SerialT m Int -> SerialT m Int)
+    -> Int
+    -> Int
+    -> SerialT m Int
+_iterateSingleton g value n = S.foldrM g (return n) $ sourceIntFromTo value n
+
+-- Apply transformation g count times on a stream of length len
+{-# INLINE iterateSource #-}
+iterateSource ::
+       S.MonadAsync m
+    => (SerialT m Int -> SerialT m Int)
+    -> Int
+    -> Int
+    -> Int
+    -> SerialT m Int
+iterateSource g count len n = f count (sourceUnfoldrM len n)
+
+    where
+
+    f (0 :: Int) stream = stream
+    f i stream = f (i - 1) (g stream)
+
+-------------------------------------------------------------------------------
+-- Functor
+-------------------------------------------------------------------------------
+
+o_n_space_functor :: Int -> [Benchmark]
+o_n_space_functor value =
+    [ bgroup "Functor"
+        [ benchIO "(+) (n times) (baseline)" $ \i0 ->
+            iterateN (\i acc -> acc >>= \n -> return $ i + n) (return i0) value
+        , benchIOSrc serially "(<$) (n times)" $
+            iterateSingleton (<$) value
+        , benchIOSrc serially "fmap (n times)" $
+            iterateSingleton (fmap . (+)) value
+        {-
+        , benchIOSrc serially "_(<$) (n times)" $
+            _iterateSingleton (<$) value
+        , benchIOSrc serially "_fmap (n times)" $
+            _iterateSingleton (fmap . (+)) value
+        -}
+        ]
+    ]
+
+-------------------------------------------------------------------------------
+-- Grouping transformations
+-------------------------------------------------------------------------------
+
+{-# INLINE groups #-}
+groups :: MonadIO m => SerialT m Int -> m ()
+groups = S.drain . S.groups FL.drain
+
+-- XXX Change this test when the order of comparison is later changed
+{-# INLINE groupsByGT #-}
+groupsByGT :: MonadIO m => SerialT m Int -> m ()
+groupsByGT = S.drain . S.groupsBy (>) FL.drain
+
+{-# INLINE groupsByEq #-}
+groupsByEq :: MonadIO m => SerialT m Int -> m ()
+groupsByEq = S.drain . S.groupsBy (==) FL.drain
+
+-- XXX Change this test when the order of comparison is later changed
+{-# INLINE groupsByRollingLT #-}
+groupsByRollingLT :: MonadIO m => SerialT m Int -> m ()
+groupsByRollingLT =
+    S.drain . S.groupsByRolling (<) FL.drain
+
+{-# INLINE groupsByRollingEq #-}
+groupsByRollingEq :: MonadIO m => SerialT m Int -> m ()
+groupsByRollingEq =
+    S.drain . S.groupsByRolling (==) FL.drain
+
+o_1_space_grouping :: Int -> [Benchmark]
+o_1_space_grouping value =
```

Comment:
Are these O(1) space? How would you represent heap proportional to group size?

Reply:
This is a mistake from my side.

Reply:
It is O(1) wrt to the stream size not the group size. If something is not O(1) it will be caught in the infinite/long stream benchmarks.